examples/ogbn_train_perforatedai.py

```diff
@@ -0,0 +1,385 @@
+
+Run docker from torch_geometric directory
+
+   docker run --gpus all -i --shm-size=8g -v .:/pai -w /pai
+   -t nvcr.io/nvidia/pyg:24.11-py3 /bin/bash
+
+Within Docker
+
+    pip install -e .
+    pip perforatedai
+    cd examples
+
+Run original with:
+
+    CUDA_VISIBLE_DEVICES=0 python ogbn_train.py --dataset ogbn-products
+    --batch_size 128
+
+Results:
+
+    Test Accuracy: 77.19%
+
+Run PAI with:
+
+    CUDA_VISIBLE_DEVICES=0 python ogbn_train_perforatedai.py --dataset
+    ogbn-products --batch_size 128 --saveName ogbnPAI
+
+Results:
+
+    Test Accuracy: 78.37%
+
+'''
+
+# Set Perforated Backpropagation settings for this training run
+'''
+# When to switch between Dendrite learning and neuron learning is determined
+by when the history of validation score or correlation scores shows those
+scores are no longer improving
+'''
+PBG.switchMode = PBG.doingHistory
+# When calculating that just look at the current score, not a recent average
+PBG.historyLookback = 1
+# How many normal epochs to wait for before switching modes.
+PBG.nEpochsToSwitch = 25
+PBG.pEpochsToSwitch = 15  # Same as above for Dendrite epochs
+'''
+The default shape of input tensors will be [batch size, number of neurons
+in the layer]
+'''
+PBG.inputDimensions = [-1, 0]
+'''
+This allows Dendrites to train as long as they keep improving rather
+than capping Dendrite training cycles to only be as long as the first
+neuron training cycle.
+'''
+PBG.capAtN = False
+# Stop the run after 4 dendrites are created
+PBG.maxDendrites = 4
+# If a set of Dendrites does not improve try 2 times before giving up
+PBG.maxDendriteTries = 2
+'''
+Make sure correlation scores improve from epoch to epoch by at least 25% and
+a raw value of 1e-4 to conclude that the correlation scores have gone up.
+'''
+PBG.pbImprovementThreshold = 0.25
+PBG.pbImprovementThresholdRaw = 1e-4
+
+parser = argparse.ArgumentParser(
+    formatter_class=argparse.ArgumentDefaultsHelpFormatter, )
+parser.add_argument(
+    '--dataset',
+    type=str,
+    default='ogbn-papers100M',
+    choices=['ogbn-papers100M', 'ogbn-products'],
+    help='Dataset name.',
+)
+parser.add_argument(
+    '--dataset_dir',
+    type=str,
+    default='./data',
+    help='Root directory of dataset.',
+)
+parser.add_argument(
+    '--use_gat',
+    action='store_true',
+    help='Whether or not to use GAT model',
+)
+parser.add_argument('-e', '--epochs', type=int, default=50000)
+parser.add_argument('--num_layers', type=int, default=3)
+parser.add_argument('--num_heads', type=int, default=2,
+                    help='number of heads for GAT model.')
+parser.add_argument('-b', '--batch_size', type=int, default=1024)
+parser.add_argument('--num_workers', type=int, default=12)
+parser.add_argument('--fan_out', type=int, default=10,
+                    help='number of neighbors in each layer')
+parser.add_argument('--hidden_channels', type=int, default=256)
+# Set to 0 to run this code without Perforated Backpropagation happening.
```

Comment:
isnt this just what the normal example does? if so i would just remove this, and update the code accordingly